CurrencyConverterControl/CurrencyConverterControl.cs

```diff
@@ -0,0 +1,346 @@
+            {
+                SetValue(InputValueProperty, value);               
+            }
+        }
+        public static readonly DependencyProperty InputValueProperty =
+            DependencyProperty.Register(nameof(InputValue), typeof(double), typeof(CurrencyConverterControl), 
+                                        new FrameworkPropertyMetadata(new PropertyChangedCallback(InputValue_Textchanged)),
+                                        validateValueCallback: new ValidateValueCallback(IsValidValue));
+
+        private static bool IsValidValue(object value)
+        {
+            double val = (double)value;
+            return !val.Equals(double.NegativeInfinity) &&
+                !val.Equals(double.PositiveInfinity);
+        }
+
+        // Adding a Text Change call back
+        private static void InputValue_Textchanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
+        {
+            try
+            {
+                CurrencyConverterControl control = (CurrencyConverterControl)d;
+                //check if all have values before going to API calculate
+                if (String.IsNullOrEmpty(control.SourceCurrency) || String.IsNullOrEmpty(control.DestinationCurrency))
+                {
+                    return;
+                }
+                control.OutputValue = CurrencyDataProvider.Convert(control.SourceCurrency,
+                                                            control.DestinationCurrency, control.InputValue);             
+              
+            }
+            catch (Exception ex) { Console.WriteLine($"Error: {ex.Message}"); }
+        }
+
+        /// <summary>
+        /// Gets or sets the output value for the result of currency conversion within the CurrencyConverterControl.
+        /// </summary>
+        /// <remarks>
+        /// The OutputValue property represents the result of the currency conversion based on the provided input value.
+        /// It is bound to the user interface elements where the converted amount is displayed to the user.
+        /// </remarks>
+        [Description("Converted value")]
+        public double OutputValue
+        {
+            get { return (double)GetValue(OutputValueProperty); }
+            set 
+            { 
+                SetValue(OutputValueProperty, value);                
+            }
+        }
+
+        public static readonly DependencyProperty OutputValueProperty =
+                            DependencyProperty.Register(nameof(OutputValue), typeof(double), typeof(CurrencyConverterControl),
+                            new FrameworkPropertyMetadata(new PropertyChangedCallback(OutputValue_Textchanged)));
+
+        private static void OutputValue_Textchanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
+        {
+            CurrencyConverterControl control = d as CurrencyConverterControl;
+            if (control != null)
+            {
+                control.OutputValue = (double)e.NewValue;
```

Comment:
Output Value is a read only textbox and it needs to be calculated based on Input Value, Currency To and Currency From. Having that in mind i didn't think it was needed to add an event on this field's changing. I added it for learning purposes, and l thought it didn't need to do anything in it. 
If l set up InputValue here and converting OutputValue into the Source Currency, l will just add extra call to the API setting a value that the end user already has written. 
Let me know if l didnt understand well.

Reply:
@rexhinahoxha we need event about outputvaluechanged. It will allow application author to perform some actions, maybe show message to end-user, do some additional validation, or update some database...

Reply:
You have added a public setter for OutputValue, so now you should handle this case. I've written about this before, that you need to either make the setter private or handle this case properly.
```
If l set up InputValue here and converting OutputValue into the Source Currency, 
l will just add extra call to the API setting a value that the end user already has written.
```
You can handle this case and avoid extra call to the API, for example, by adding a flag and not calling the conversion (check this flag) if it is not necessary.